File: src/handlers/auth/signUp.ts
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import middy from '@middy/core';
import httpJsonBodyParser from '@middy/http-json-body-parser';
import httpErrorHandler from '@middy/http-error-handler';
import {
  CognitoIdentityProviderClient,
  SignUpCommand,
  UsernameExistsException,
  InvalidPasswordException,
} from '@aws-sdk/client-cognito-identity-provider';
import { z } from 'zod';
import { logger } from '../../shared/powertools';
import { config } from '../../shared/config';

const cognito = new CognitoIdentityProviderClient({});

const SignUpSchema = z.object({
  email: z.string().email(),
  password: z.string().min(8),
  name: z.string().optional(),
});

const lambdaHandler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
  const parseResult = SignUpSchema.safeParse(event.body);
  if (!parseResult.success) {
    return res(400, { message: parseResult.error.issues[0].message });
  }

  const { email, password, name } = parseResult.data;
  logger.info('Sign up attempt', { email });

  try {
    await cognito.send(new SignUpCommand({
      ClientId: config.userPoolClientId,
      Username: email,
      Password: password,
      UserAttributes: [
        { Name: 'email', Value: email },
        ...(name ? [{ Name: 'name', Value: name }] : []),
      ],
    }));

    logger.info('Sign up successful', { email });
    return res(200, { message: 'Sign up successful. Check your email for a verification code.' });
  } catch (err) {
    if (err instanceof UsernameExistsException) {
      return res(409, { message: 'An account with this email already exists.' });
    }
    if (err instanceof InvalidPasswordException) {
      return res(400, { message: (err as Error).message });
    }
    logger.error('Sign up failed', { email, error: String(err) });
    throw err;
  }
};

export const handler = middy(lambdaHandler)
  .use(httpJsonBodyParser())
  .use(httpErrorHandler());

function res(statusCode: number, body: object): APIGatewayProxyResult {
  return {
    statusCode,
    headers: { 'Access-Control-Allow-Origin': '*', 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  };
}

File: src/handlers/auth/sign_up.py
import json
import re

import boto3
from botocore.exceptions import ClientError

from src.shared.config import config
from src.shared.powertools import logger, tracer

cognito = boto3.client("cognito-idp")

_EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")


def _res(status_code: int, body: dict) -> dict:
    return {
        "statusCode": status_code,
        "headers": {"Access-Control-Allow-Origin": "*", "Content-Type": "application/json"},
        "body": json.dumps(body),
    }


@logger.inject_lambda_context
@tracer.capture_lambda_handler(capture_response=False)
def handler(event, context):
    try:
        body = json.loads(event.get("body") or "{}")
    except (json.JSONDecodeError, TypeError):
        return _res(400, {"message": "Invalid JSON body"})

    email = str(body.get("email", "")).strip()
    password = str(body.get("password", ""))
    name = str(body.get("name", "")).strip()

    if not email or not _EMAIL_RE.match(email):
        return _res(400, {"message": "Invalid email address"})
    if len(password) < 8:
        return _res(400, {"message": "Password must be at least 8 characters"})

    logger.info("Sign up attempt", email=email)

    user_attributes = [{"Name": "email", "Value": email}]
    if name:
        user_attributes.append({"Name": "name", "Value": name})

    try:
        cognito.sign_up(
            ClientId=config["user_pool_client_id"],
            Username=email,
            Password=password,
            UserAttributes=user_attributes,
        )
        logger.info("Sign up successful", email=email)
        return _res(200, {"message": "Sign up successful. Check your email for a verification code."})
    except ClientError as e:
        error_code = e.response["Error"]["Code"]
        if error_code == "UsernameExistsException":
            return _res(409, {"message": "An account with this email already exists."})
        if error_code == "InvalidPasswordException":
            return _res(400, {"message": e.response["Error"]["Message"]})
        logger.error("Sign up failed", email=email, error=str(e))
        raise
